### ISSUE TITLE
A timeout in Mocha before() step causes `wdio` to exit uncleanly.

### ISSUE THREAD

**Issue by tim-mit** [webdriverio/webdriverio#2259]
## The problem

When a timeout occurs in a Mocha `before()` step `wdio` exits uncleanly and doesn't report what error actually occurred.

This looks to be closely related to https://github.com/webdriverio/webdriverio/issues/1460 but that issue is marked as closed. There was a fix merged for that ticket, https://github.com/webdriverio/webdriverio/commit/305ed2c05dafff58041a3e6f4804042ffcbc40cf but it was rolled back a couple of days later, https://github.com/webdriverio/webdriverio/commit/27e825237325ef5f624c41e11c65815ef9e777c2.

Note that a timeout within a `describe()` block is correctly reported on and the reason for the test is logged.

## Environment

* WebdriverIO version: 4.8.0
* Node.js version: v6.3.1

* [Standalone mode or wdio testrunner](http://webdriver.io/guide/getstarted/modes.html): wdio
* if wdio testrunner, running synchronous or asynchronous tests: sync
* Additional wdio packages used (if applicable):
```
"wdio-mocha-framework": "^0.5.10",
"wdio-selenium-standalone-service": "0.0.9",
"wdio-spec-reporter": "^0.1.2",
```

## Details

I've created an example test suite that shows the problem at https://github.com/tim-mit/wdio-test-suite. The test has a `before` step that loads a url, in this case www.google.com, and then waits for an element that times out which simulated by waiting for an id that doesn't exist in my test.

Instead of getting the actual error (that `browser.waitForVisible` call timed out) an internal error within wdio `ReporterStats.js` is logged.

## WebdriverIO logs

An example output;

```sh
$ npm run wdio

> wdio-test-suite@0.0.1 wdio /Users/tim-mit/wdio-test-suite
> wdio wdio.conf.js

/Users/tim-mit/wdio-test-suite/node_modules/webdriverio/build/lib/utils/ReporterStats.js:452
            testStats.state = 'fail';
                            ^

TypeError: Cannot set property 'state' of undefined
    at ReporterStats.testFail (/Users/tim-mit/wdio-test-suite/node_modules/webdriverio/build/lib/utils/ReporterStats.js:452:29)
    at BaseReporter.<anonymous> (/Users/tim-mit/wdio-test-suite/node_modules/webdriverio/build/lib/utils/BaseReporter.js:159:25)
    at emitOne (events.js:96:13)
    at BaseReporter.emit (events.js:188:7)
    at BaseReporter.handleEvent (/Users/tim-mit/wdio-test-suite/node_modules/webdriverio/build/lib/utils/BaseReporter.js:300:27)
    at Launcher.messageHandler (/Users/tim-mit/wdio-test-suite/node_modules/webdriverio/build/lib/launcher.js:623:28)
    at emitTwo (events.js:106:13)
    at ChildProcess.emit (events.js:191:7)
    at process.nextTick (internal/child_process.js:719:12)
    at _combinedTickCallback (internal/process/next_tick.js:67:7)
```

## Code To Reproduce Issue
Example test that shows the issue is available at https://github.com/tim-mit/wdio-test-suite.

## Workaround (potential fix?)

I've taken a look at the code and managed to fix the issue for myself but don't know the knock on effects of the change I've made so haven't created a PR. In [`ReporterStats.js`](https://github.com/webdriverio/webdriverio/blob/master/lib/utils/ReporterStats.js#L310) I've made sure that the call to `testStats = this.getTestStats(runner)` always assigns an object to `testStats` regardless of whether or not the test fails by including ` || {}` on the end of the execution - not the prettiest code but it seems to work, perhaps it'd be tidier to return an object directly at line [232](https://github.com/webdriverio/webdriverio/blob/master/lib/utils/ReporterStats.js#L232)?

```js
              try {
                  testStats = this.getTestStats(runner) || {};
              } catch (e) {
                  // If a test fails during the before() or beforeEach() hook, it will not yet
                  // have been 'started', so start now
                  this.testStart(runner);
                  testStats = this.getTestStats(runner);
              }
```

With this in place I get a 'good' failure output that tells me that the test failed due to a timeout;

```sh
$ npm run wdio

> wdio-test-suite@0.0.1 wdio /Users/tim-mit/wdio-test-suite
> wdio wdio.conf.js

------------------------------------------------------------------
[chrome #0-0] Session ID: 04a4c56ca01b19031b5423040b618b80
[chrome #0-0] Spec: /Users/timmit/wdio-test-suite/test.spec.js
[chrome #0-0] Running: chrome
[chrome #0-0]
[chrome #0-0] Google Test
[chrome #0-0]
[chrome #0-0]
[chrome #0-0] 1 failing (19s)
[chrome #0-0]
[chrome #0-0] 1) Google Test "before all" hook:
[chrome #0-0] element ("#not-an-id-on-the-page") still not visible after 10000ms
[chrome #0-0] Error: element ("#not-an-id-on-the-page") still not visible after 10000ms
[chrome #0-0]     at Context.<anonymous> (test.spec.js:9:11)
[chrome #0-0]     at new Promise (node_modules/core-js/library/modules/es6.promise.js:177:7)
[chrome #0-0]     at elements("#not-an-id-on-the-page") - isVisible.js:54:17
[chrome #0-0]     at isVisible("#not-an-id-on-the-page") - waitForVisible.js:73:22
[chrome #0-0]
```

**Comment by aurelhann** [comment]
Same Issue in selenium standalone mode with mocha....

**Comment by tim-mit** [comment]
I'm happy to submit a PR for this with the fix I've been using locally if that's wanted. The fix seems to work fine but I don't want to just propose a band-aid solution if without someone with experience of the code in question being able to tell me it needs to be handled in a cleaner way.

**Comment by christian-bromann** [comment]
@tim-mit you should not have the before hook outside of the describe block. Feel free to submit a PR to fix this but test hooks should be within describe blocks

**Comment by christian-bromann** [comment]
Can not reproduce this. Please don't use hooks before the first describe block. This would fail:
```js
before(() => {
  // ..
})

describe('test' , () => {
  // ...
})
```
Please do:

```js
describe('test' , () => {
  before(() => {
    // ..
  })
  // ...
})
```

**Comment by christian-bromann** [comment]
I found a fix for this issue that will allow to do both. It will be released with `v4.9.11`.
